get-poetry.py

```diff
@@ -187,7 +187,11 @@ def expanduser(path):
 
 
 HOME = expanduser("~")
-POETRY_HOME = os.environ.get("POETRY_HOME") or os.path.join(HOME, ".poetry")
+POETRY_HOME = os.getenv("POETRY_HOME", os.getenv("XDG_PACKAGE_HOME"))
```

Comment:
This could lead to unexpected behavior if `POETRY_HOME` is not a legal or existing path.

```
>>> import os
>>> os.path.join("not a legal path", "/home")
'/home'
```

Reply:
Interesting! Let me biff up this code on the next patch.

Reply:
Had some though over it and this is what I come up with:

```python
for target_dir in [os.getenv("POETRY_HOME"), os.getenv("XDG_PACKAGE_HOME")]:
    if target_dir and os.path.exists(os.path.dirname(expanduser(target_dir))):
        POETRY_HOME = expanduser(target_dir)
        break
else:
    POETRY_HOME = os.path.join(HOME, ".poetry")
```

I wanted to use a list comprehension but doing so would make all 3 possibilities to be evaluated. With a for loop, we only evaluate the paths up to the first that has a valid parent directory (e.g. `"not a legal path"` returns a `False` on the check, but `"/not a legal path"` or "~/not a legal path" are perfectly valid paths.

Reply:
on a second thought, would be acceptable to use: `os.path.join(target_dir, "test") != "test"` as the if clause?

Reply:
After playing a bit with this code I found that the original solution would still works. The problem with the the example:

```python
>>> import os
>>> os.path.join("not a legal path", "/home")
'/home'
```

Is the `/` at the beginning of the second argument of `join`. Since `join` is meant to concatenate paths, when you define a full path as one of the arguments, `join` will hapily return that directory as your root:

```python
>>> import os
>>> os.path.join("/valid/path", "/home")
'/home'
```

Now, when you restrict the second argument without the initial `/`, even the first example works:

```python
>>> import os
>>> os.path.join("not a legal path", "home")
'not a legal path/home'
```

Since the main purpose of `get-poetry.py` is to install a new software on an unexistent path, I'll rollback to my initial change.